gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineJdkRelease.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.plugins;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.gradle.api.JavaVersion;
+import org.gradle.api.Project;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+/**
+ * <p>
+ * When using JDK 9+ to compile with a targetCompatibility less than JDK 9, this plugin adds compiler arguments
+ * per <a href="https://openjdk.java.net/jeps/247">JEP 247</a> to explicitly set the target JDK platform API
+ * to maintain binary compatibility.
+ * </p>
+ * <p>
+ * See also <a href="https://github.com/gradle/gradle/issues/2510">Gradle JDK release issue</a>.
+ * </p>
+ */
+public final class BaselineJdkRelease extends AbstractBaselinePlugin {
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        project.getPluginManager().withPlugin("java", plugin -> {
+            project.getTasks().withType(JavaCompile.class).configureEach(javaCompile -> {
+                try {
+                    configureJavaCompiler(javaCompile);
+                } catch (Throwable throwable) {
+                    project.getLogger().error("Failed to configure BaselineJdkRelease", throwable);
+                }
+            });
+        });
+    }
+
+    private void configureJavaCompiler(JavaCompile javaCompile) {
+        JavaVersion sourceCompatibility = JavaVersion.toVersion(javaCompile.getTargetCompatibility());
```

Comment:
source vs target, is that on purpose? if so, please comment and explain

Reply:
I think we want use `targetCompatibility` in general (at least as default), though if we wanted to allow flexibility we could introduce a separate `releaseCompatibility` extension property.

* `sourceCompatibility`: The version of Java source code language features that are permitted by the compiler. This must be less than or equal to `target`
* `targetCompatibility`: The version of JVM bytecode class files that will be generated. This must be greater than or equal to `source`
* `releaseCompatibility`: The version of JDK API libraries to link against (equivalent to pre JDK 9 `bootclasspath`). This implicitly indicates that `source == target == release`

See also https://docs.oracle.com/en/java/javase/11/tools/javac.html#GUID-AEEC9F07-CB49-4E96-8BC7-BCC2C7F725C9 and https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javac.html#BHCIJIEG 

Note that you can't use `source` and `target` with `release` for `javac`:
```
$ javac -version -source 6 -target 7 --release 8 Foo.java
javac 11.0.2
error: option -source cannot be used together with --release
error: option -target cannot be used together with --release
```